# Freshwater_sponge_microbiome

This repository provides supporting information for the manuscript "Microbiome of the freshwater sponge <i>Ephydatia muelleri</i> shares compositional and functional similarities with those of marine sponges," which has been submitted for peer review.

- <b>sp_analysis</b> provides the R code to reproduce all analyses and statistical tests.
- <b>sp_figures</b> provides the R code to reproduce all figures. It requires objects produced by the analysis code.
- <b>sp_qiime2_code</b> provides the Linux terminal code used for all QIIME2 analysis (i.e., comparison to Kenny et al.).
- <b>sp_other_code</b> provides the Linux terminal code (or examples of the code) used for MAFFT alignment and MAG analyis. 

The raw data will be provided as an R workspace and associated files upon manuscript acceptance or upon reasonable request to the first author, who can be reached at sasugden@gmail.com or ssugden@ualberta.ca.

If you have questions or concerns with these files, or trouble downloading them, please contact sasugden@gmail.com or ssugden@ualberta.ca.

De-multiplexed, unfiltered sequencing data has been deposited in the NCBI Short Read Archive under accession number PRJNA526747.

This workspace was last updated on April 1, 2022.
